### PR TITLE
Dl2

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

File: src/cryptonote_core/cryptonote_core.cpp [hunk 1]
```diff
@@ -134,6 +135,14 @@ namespace cryptonote
   void core::stop()
   {
     m_blockchain_storage.cancel();
+
+    boost::lock_guard<boost::mutex> lock(m_update_mutex);
+    tools::download_async_handle handle = m_update_download;
```
Comment (vtnerd) [inline src/cryptonote_core/cryptonote_core.cpp]: Ok, just saw your comment in the other PR in my inbox. I didn't even notice the data race at first actually. Its still a race because the callback could be invoked right as this code block is running.

The fix is to lock, copy handle, release, then call cancel. The callback would also lock.

File: src/common/download.h [hunk 2]
```diff
@@ -32,5 +32,13 @@
 
 namespace tools
 {
-  bool download(const std::string &path, const std::string &url);
+  struct download_thread_control;
+  typedef std::shared_ptr<download_thread_control> download_async_handle;
+
+  bool download(const std::string &path, const std::string &url, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress = NULL);
+  download_async_handle download_async(const std::string &path, const std::string &url, std::function<void(const std::string&, const std::string&, bool)> result, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress = NULL);
+  bool download_error(download_async_handle h);
```
Comment (vtnerd) [inline src/common/download.h]: Also, a `const download_async_handle&` signature prevents an atomic ref count when calling the function.

File: src/cryptonote_core/cryptonote_core.cpp [hunk 3]
```diff
@@ -1133,32 +1142,64 @@ namespace cryptonote
...
+            {
+              MCERROR("updates", "Download from " << uri << " does not match the expected hash");
+            }
+            MCLOG_CYAN(el::Level::Info, "updates", "New version downloaded to " << path);
+          }
+          else
+          {
+            MCERROR("updates", "Failed to download " << uri);
+          }
+          m_update_download = 0;
```
Comment (vtnerd) [inline src/cryptonote_core/cryptonote_core.cpp]: Crap, a destructor would have to lock because the callback modifies the value. Well, this stinks either way. The likelihood of stop being skipped right is now is low at least.

File: src/common/download.cpp [hunk 4]
```diff
@@ -40,27 +41,96 @@
...
+
+    control->started_lock.lock();
+    control->started = true;
+    control->started_cond.notify_one();
+    control->started_lock.unlock();
+
     try
     {
-      MINFO("Downloading " << url << " to " << path);
+      unlockable_lock_guard lock(control->mutex);
```
Comment (vtnerd) [inline src/common/download.cpp]: Ok, I'm sorry for being so irritating but I just remembered its 100% easier to do this in the destructor of `~download_thread_control() { if (thread.joinable()) thread.detach(); }`. Also, the way I suggested would be bad because the lock is not held on the `thread.join()` calls so pthread would get mad.

And perhaps comments should be added for the exposed functions doing the join calls - it is not safe to call those functions from two different threads.

File: src/common/download.cpp [hunk 5]
```diff
@@ -40,27 +41,96 @@
...
+
+  struct download_thread_control
+  {
+    const std::string path;
+    const std::string uri;
+    std::function<void(const std::string&, const std::string&, bool)> result_cb;
+    std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb;
+    bool started;
+    bool stop;
+    bool stopped;
```
Comment (moneromooo-monero) [inline src/common/download.cpp]: Though I don't see a is_joinable or similar in the class definition.

Comment (moneromooo-monero) [inline src/common/download.cpp]: nvm, I was looking at an old versoin of the class

**Comment by moneromooo-monero** [comment]
A separate process only helps to recover in case of crash, not against exploitation (unless you go further than that). You'd need to at least chown to a different user (I guess nobody does the trick for Linux though). That seems like a lot of trouble for not much, as you'd need RPC between the two.

**Comment by vtnerd** [comment]
The possibility of a use-after-free just had me thinking about process separation for memory errors. A stray pointer could affect a portion of the database which is mapped into memory as writeable. Also, a separate process could also aggressively disable the creation of new file descriptors after opening the socket and file. Lots of work for a small bump in protection.

File: src/common/download.cpp [hunk 6]
```diff
@@ -40,27 +41,76 @@
...
+    std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb;
+    bool stop;
+    bool success;
+    boost::thread thread;
+    boost::mutex mutex;
+
+    download_thread_control(const std::string &path, const std::string &uri, std::function<void(const std::string&, const std::string&, bool)> result_cb, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb):
+        path(path), uri(uri), result_cb(result_cb), progress_cb(progress_cb), stop(false), success(false) {}
+    ~download_thread_control() { if (thread.joinable()) thread.detach(); }
+    void lock() { mutex.lock(); }
```
Comment (vtnerd) [inline src/common/download.cpp]: BTW, I think these are unused now.

File: src/common/download.cpp [hunk 7]
```diff
@@ -70,69 +120,145 @@ namespace tools
...
+    download_async_handle control = std::make_shared<download_thread_control>(path, url, result, progress);
+    control->thread = boost::thread([control](){ download_thread(control); });
+    return control;
+  }
+
+  bool download_finished(const download_async_handle &control)
+  {
+    CHECK_AND_ASSERT_MES(control != 0, false, "NULL async download handle");
+    boost::lock_guard<boost::mutex> lock(control->mutex);
+    return !control->thread.joinable();
```
Comment (vtnerd) [inline src/common/download.cpp]: And this doesn't work anymore. `joinable()` remains true until someone detaches or joins. Since I told you to move the detach call from the `download_thread` function, this will only be true if someone calls `download_error` or `download_cancel` first.

So bad idea.

**Comment by moneromooo-monero** [comment]
I only detach from the async handle dtor.

**Comment by hyc** [comment]
Lee Clagett wrote:
> The possibility of a use-after-free just had me thinking about process
> separation for memory errors. A stray pointer could affect a portion of the
> database which is mapped into memory as writeable. Also, a separate process
> could also aggressively disable the creation of new file descriptors after
> opening the socket and file. Lots of work for a small bump in protection.

Keep in mind that by default the DB mmap is read-only. In "fastest" sync mode 
it is writable, but not in safe or fast mode.

-- 
   -- Howard Chu
   CTO, Symas Corp.           http://www.symas.com
   Director, Highland Sun     http://highlandsun.com/hyc/
   Chief Architect, OpenLDAP  http://www.openldap.org/project/

**Comment by vtnerd** [comment]
> Keep in mind that by default the DB mmap is read-only. In "fastest" sync mode it is writable, but not in safe or fast mode.

Thanks for pointing this out, I actually did not realize this. This "mixed mode" - file write functions with mmap reads - is an interesting security/integrity/performance compromise.
